날씨 앱 [STEP3] 요한

dpwns1234 opened the pull request:
# 날씨 앱 [STEP3] 요한
리뷰어: @coryKim0829

안녕하세요 코리! 요한입니다.
STEP3 PR 보냅니다. 잘 부탁드립니다-!!


## 구현 영상


![화면 기록 2023-12-09 오전 2 16 50](https://github.com/tasty-code/ios-weather-forecast/assets/52391722/2b485f44-a8d4-4d67-8a74-667f39bee4a3)


## 파일구조📁
- Secrets.xcconfig
- Controller
    - WeatherViewController
- Data
    - WeatherDataManagerDelegate
    - WeatherDataManager
    - Location
        - LocationDataManagerDelegate
        - LocationDataManager
- Extensoins
    - Double
- Error
    - NetworkError
- Location
    - LocationDataManagerDelegate
    - LocationDataManager
- Model
    - WeatherToday
    - WeatherForecast
    - WeatherCommonInformation
- Network
    - WeatherNetworkManagerDelegate
    - WeatherNetworkManager
    - ImageCacheManager
- Service
    - ServiceType
    - DataServiceable
    - ForecastDataServiceDelegate
    - ForecastDataService
    - TodayDataServiceDelegate
    - TodayDataService
    - IconDataServiceDelegate
    - IconDataService
    - 
- View
    - WeatherCollectionViewCell
    - WeatherCollectionHeaderView



## 구현내용

### 구현 사항

- [x] 앱의 UI를 구현합니다
    - [x] 오토레이아웃을 활용하여 코드로만 구현
    - [x] Collection View를 활용
    - [ ]  Configuration을 활용
- [x] 화면을 위에서 아래로 끌었다 놓으면 정보를 새로고침합니다.


## 트러블 슈팅
### 문제
기존 NetworkManager를 통해 받아오는 구조가 decodable한 객체를 받아올 수 밖에 없는 코드였고, 날씨 icon을 받아올 수 없었습니다. icon image를 받아오려면 새로 URLSession을 통해서야 받아올 수 있었습니다. 
결국 날씨 예보, 현재 날씨 그리고 날씨 아이콘 모두를 받아올 수 있는 NetworkManager를 구현해야 한다 생각했고, 그 과정에서 여러 트러블이 발생했습니다.

### 해결
먼저 데이터 타입에 따라 Service클래스를 만들고
-> `ForecastDataService`, `TodayDataService`, `IconDataService`
이 Service안에 NetworkManager를 두어 데이터를 download 및 데이터 타입으로 변환하도록 설계하였습니다. 이를 통해 한 곳에서 네트워크 통신을 하도록 구현하였습니다.

또한 추후에 새로운 데이터 타입의 변환이 필요할 경우 수정에 용이하도록 DataServiceable 프로토콜을 채택하도록 구현하였습니다.



```swift 
// DataServiceable.swift

protocol DataServiceable: AnyObject {
    var decoder: JSONDecoder { get }
    var networkManager: WeatherNetworkManager { get }
    
    func downloadData(type service: ServiceType) throws
}

extension DataServiceable {
    var decoder: JSONDecoder {
        return JSONDecoder()
    }
    
    var networkManager: WeatherNetworkManager {
        return WeatherNetworkManager()
    }
}

// ForecastDataService.swift

final class ForecastDataService: DataServiceable {
    weak var delegate: ForecastDataServiceDelegate?
    
    func downloadData(type service: ServiceType) throws {
        guard let url = service.makeURL() else { throw NetworkError.invailedURL }
        ...
    }
}

// TodayDataService.swift

final class TodayDataService: DataServiceable {
    weak var delegate: TodayDataServiceDelegate?
    
    func downloadData(type service: ServiceType) throws {
        guard let url = service.makeURL() else { throw NetworkError.invailedURL }
        ...
    }
}

// IconDataService.swift

final class IconDataService: DataServiceable {
    weak var delegate: IconDataServiceDelegate?
    
    func downloadData(type service: ServiceType) throws {
        guard let url = service.makeURL() else { throw NetworkError.invailedURL }
        ...
    }
}
```


## 궁금한 점
현재 collecionView의 header와 cell의 데이터 바인딩을 ViewController에서 해주고 있습니다. ViewController 공식문서의 "Updating the contents of the views, usually in response to changes to the underlying data" 를 근거로 viewController에서 binding을 해주고 있는데 VC의 코드가 길어지기도 하고, 각 header 또는 cell 안에서 해주는 게 더 깔끔한 느낌도 있습니다. 혹시 코리는 어떻게 생각하는지 궁금합니다!

dpwns1234 committed:
refactor: delegate 함수명 명확하게 변경

dpwns1234 committed:
feat: Header UI 구현

dpwns1234 committed:
feat: CollectionView custom cell 클래스 구현

dpwns1234 committed:
refactor: network -> viewController 의 데이터 전달 방법 변경

- delegate 가 아닌 cloure를 이용해 데이터 전달
- forecast, todayWeather 모두를 받아오는 과정에서 로직 오류로 인해
NetworkManager를 singleton 패턴으로 적용

dpwns1234 committed:
feat: collectoinView cell binding 및 UI 수정

dpwns1234 committed:
refactor: collectionView header 제약조건 수정

dpwns1234 committed:
feat: 배경 image 추가

dpwns1234 committed:
feat: refreshControl 추가 및 collectionView reload() 추가

- 데이터를 가져올 시점이 network에서 돌아가던 것이기에 main이 아닌 스레드에서 작동
그렇기에 main 스레드에서 reload를 해준다.

dpwns1234 committed:
fix: main에 잘못 추가한 label 삭제

dpwns1234 committed:
fix: cell 스크롤 시 재사용으로 계절 이미지 아이콘이 바뀌는 문제 해결

dpwns1234 committed:
refactor: viewController UI refactor

- ui 설정하는 코드를 한 곳으로 몰아넣음
- yongsan 이미지 이름 변경

dpwns1234 committed:
refactor: bind 위치를 viewController 로 이동

- viewController의 역할 중 bind가 있음

dpwns1234 committed:
refactor: weatherApiClient 리팩터

- weatherApiClien와 weatherType을 enum type으로 합침

dpwns1234 committed:
refactor: weahterApiClient 네이밍 변경

- enum 타입이며, dataService의 타입을 결정해주므로 적절한 ServiceType으로 이름 변경

dpwns1234 committed:
feat: decodable 타입 객체를 decode할 클래스 분리

dpwns1234 committed:
fix: components 오타 수정

dpwns1234 committed:
feat: error 타입 추가

dpwns1234 committed:
refactor: 새로운 weatherNetworkManager에 맞게 코드 리팩터

dpwns1234 committed:
refactor: weatherNetworkManager 수정

dpwns1234 committed:
chore: 해당 클래스 이름에 맞게 주석 수정

dpwns1234 committed:
refactor: DataService의 확장성을 위한 DataService 프로토콜 구현

dpwns1234 committed:
chore: 필요없는 코드 삭제

dpwns1234 committed:
chore: header, cell cutomView View group으로 이동

dpwns1234 committed:
feat: cache 설정

- icon을 미리 저장해두고 사용하려고.
- 한 번에 로드되려면 cache가 필요하다고 판단

dpwns1234 committed:
feat: iconData를 제공하는 IconDataService 클래스 구현

- cacheManager를 사용해 download 동시에 cache에 저장

dpwns1234 committed:
refactor: 프로퍼티 및 메서드 namaing 수정

dpwns1234 committed:
refactor: WeatherDataManagerDelegate 메스드명 변경

dpwns1234 committed:
refactor: 데이터를 받아오는 방식의 리팩터에 따른 viewController 리팩터

dpwns1234 committed:
refactor: 섭씨 온도 변환 및 string 변환 함수 extension으로 리팩터

dpwns1234 committed:
refactor: locationDataManager을 DataManager에서 관리하도록 리팩터

dpwns1234 committed:
chore: 필요없는 코드 삭제

dpwns1234 committed:
refactor: error type 추가

dpwns1234 committed:
chore: .pbxproj

dpwns1234 committed:
chore: 주석 제거

dpwns1234 committed:
chore: 파일 그룹 정리

dpwns1234 committed:
design: cell constans 수정

dpwns1234 committed:
chore: 필요없는 파일 삭제

dpwns1234 committed:
design: 배경이미지 변경 및 ui constants 변경

corykim0829 commented on WeatherForecast/WeatherForecast/Network/WeatherNetworkManager.swift:
```diff
@@ -1,73 +1,33 @@
-            } catch {
-                print(error)
+final class WeatherNetworkManager {
+    func downloadData(url: URL, completionHandler: @escaping (Result<Data, Error>) -> Void) {
```
Result 타입을 사용해서 completionHandler 클로저를 잘 구현하셨네요!
예외 처리도 모두 잡아준 부분이 좋습니다 👍

corykim0829 commented on WeatherForecast/WeatherForecast/Data/WeatherDataManager.swift:
```diff
@@ -0,0 +1,93 @@
+    weak var delegate: WeatherDataManagerDelegate?
+    
+    // MARK: Private property
+    private let locationDataManager = LocationDataManager()
```
WeatherDataManager가 locationManager까지 가지고 있는다면, WeatherData가 하는 일이 너무 많아지고 복잡해지는 느낌이 있네요.
LocationManager는 ViewController에서 따로 처리하는건 어떨까요?

dpwns1234 replied:
비슷한 역할을 하는 코드들은 한 클래스로 모아야 한다고 생각했는데, 그렇게 함으로 한 클래스에서 너무 많은 일을 하게 되는 부분은 생각하지 못했던 것 같습니다. LocationManager를 VC로 옮겨 따로 처리하도록 리팩터 하였습니다!

corykim0829 commented on WeatherForecast/WeatherForecast/Controller/WeatherViewController.swift:
```diff
@@ -68,10 +169,54 @@ extension WeatherViewController: LocationDataManagerDelegate {
+    }
+    
+    @objc
+    private func handleRefreshControl() {
```
제가 생각했을 때에는 여기서 collectionView만 reload하면 실질적인 데이터는 바뀌지 않을 것 같네요.
데이터를 새롭게 요청하는 코드가 추가로 필요해보입니다!

dpwns1234 replied:
헉.. 바로 수정했습니다!

corykim0829 commented on WeatherForecast/WeatherForecast/Controller/WeatherViewController.swift:
```diff
@@ -68,10 +169,54 @@ extension WeatherViewController: LocationDataManagerDelegate {
-    func weather<T>(_ manager: WeatherNetworkManager, didLoad data: T) where T : Decodable {
-        print(data)
+extension WeatherViewController {
+    private func bind(on header: WeatherCollectionHeaderView) {
```
궁금하신 점에서 말씀하신 것 처럼, ViewController에서는 데이터를 뷰에게 전달만 해주고 뷰단에서 뷰를 업데이트하는 코드를 구현하는게 더 깔끔하고 좋습니다!

corykim0829 commented on WeatherForecast/WeatherForecast/Controller/WeatherViewController.swift:
```diff
@@ -6,41 +6,142 @@
+
+extension WeatherViewController: UICollectionViewDelegateFlowLayout {
+    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
+        return CGSize(width: view.frame.width, height: 60)
```
매직넘버를 상수로 정리하면 좋을 것 같습니다!

dpwns1234 replied:
enum type으로 정의해서 정리 후 커밋 올렸습니다!

corykim0829 commented on WeatherForecast/WeatherForecast/Controller/WeatherViewController.swift:
```diff
@@ -6,41 +6,142 @@
+    }
+    
+    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, referenceSizeForHeaderInSection section: Int) -> CGSize {
+        return CGSize(width: view.frame.size.width, height: 120)
```
여기도 매직넘버 정리하면 좋을 것 같습니다!

dpwns1234 replied:
어떤 데이터가 로드 되었든, 같은 역할을 하기에 하나의 함수를 중복해서 호출해주었습니다.
그런데 뷰컨에게 어떤 데이터가 언제 들어왔는지 업데이트 시점을 알려주는 것을 생각하지 못했습니다.. 
5953f28dcea52419ece95de7475e38d0c9c9b48f 해당 커밋에서 둘로 분리하였는데 코리의 의견에 맞게 변경하였을까요?

corykim0829 commented on WeatherForecast/WeatherForecast/Controller/WeatherViewController.swift:
```diff
@@ -6,41 +6,142 @@
         super.viewDidLoad()
-        locationDataManager.locationDelegate = self
-        weatherNetworkManager.weatherDelegate = self
+        setUI()
+        setRefreshControl()
```
함수로 코드들을 분리시킨 부분이 깔끔하고 좋습니다!

corykim0829 commented on WeatherForecast/WeatherForecast/Controller/WeatherViewController.swift:
```diff
@@ -6,41 +6,142 @@
+
+extension WeatherViewController: UICollectionViewDataSource {
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        return dataManager.forecast?.list.count ?? 20
```
20이 왜 사용된건지 코드만 보면 이해가 잘 되지 않습니다.
또한 해당값이 없을 때 옵셔널 처리를 한번 더 고민해보면 좋을 것 같습니다!

dpwns1234 replied:
임시적으로 20으로 처리해놓고 따로 수정을 못했습니다..
list.count가 nil이라는 것은 아직 data가 로드되지 않았다는 것으로 ?? 0 으로 수정하였는데,
따로 옵셔널 처리하는 것은 어떻게 처리를 해야하는지 잘 모르겠습니다..

corykim0829 commented on WeatherForecast/WeatherForecast/Controller/WeatherViewController.swift:
```diff
@@ -6,41 +6,142 @@
+// MARK: - WeatherDataManagerDelegate
+
+extension WeatherViewController: WeatherDataManagerDelegate {
+    func completedLoadData(_ manager: WeatherDataManager) {
```
이 함수가 두가지 데이터를 받을 때 어떤 데이터가 정확하게 언제 들어왔는지 알 수 없는 것 같아요.
그래서 제가 보기에는 중복으로 호출하는 느낌이 드네요.
분리해서 뷰컨트롤러에게 업데이트 시점을 알려주는게 좋을 것 같아요

dpwns1234 replied:
어떤 데이터가 로드 되었든, 같은 역할을 하기에 하나의 함수를 중복해서 호출해주었습니다.
그런데 뷰컨에게 어떤 데이터가 언제 들어왔는지 업데이트 시점을 알려주는 것을 생각하지 못했습니다.. 
5953f28dcea52419ece95de7475e38d0c9c9b48f 해당 커밋에서 둘로 분리하였는데 코리의 의견에 맞게 변경하였을까요?

corykim0829 replied:
네 잘 수정하셨습니다!

corykim0829 commented on WeatherForecast/WeatherForecast/Controller/WeatherViewController.swift:
```diff
@@ -6,41 +6,142 @@
+// MARK: - Set ViewController UI
+
+extension WeatherViewController {
+    private func setUI() {
```
만약 하나의 뷰컨트롤러에 UI 요소가 많아진다면, 이 함수가 매우 커지게 됩니다.
그렇기 때문에 각 뷰의 구현을 함수로 분리하거나 커스텀 클래스로 구현하면 더 깔끔하니 참고만하세요!

corykim0829 commented on WeatherForecast/WeatherForecast/Controller/WeatherViewController.swift:
```diff
@@ -6,41 +6,142 @@
+        setConstraint()
+    }
+    
+    private func setConstraint() {
```
오토레이아웃도 구현을 잘 하셨습니다!
UI요소와 마찬가지로 이 부분도 함수로 분리하면 더 가독성이 좋으니 참고하세요!

corykim0829 commented on WeatherForecast/WeatherForecast/Service/IconDataService.swift:
```diff
@@ -0,0 +1,30 @@
+                    print("이미지 변환에 실패했습니다.")
+                    return
+                }
+                ImageCacheManager.setCache(image: image, forKey: service.code)
```
이미지 캐싱까지 구현하신 부분 좋습니다! 👍

corykim0829 commented on WeatherForecast/WeatherForecast/Service/IconDataService.swift:
```diff
@@ -0,0 +1,30 @@
+                ImageCacheManager.setCache(image: image, forKey: service.code)
+                self.delegate?.didCompleteLoad(self)
+            case .failure(let error):
+                print(error)
```
이미지 데이터를 가져오지 못했을 때에는 처리가 따로 되어있지 않았네요
delegate는 두개의 함수를 만들어야하고 delegate 대상을 지정해야하는 번거로움이 조금 있으니 delegate 보다 클로저 파라미터를 이용해서 Icon 데이터 가져오는 것을 처리하는게 어떨까요?

corykim0829 reviewed (changes requested):
수고하셨습니다!

WeatherDataManager가 현재는 핵심 로직을 다 가지고 있는데, 이를 VC에서 컨트롤하는 부분을 캐치하기 힘든 것 같습니다.

또한 delegate를 잘 사용하셨지만 여러 delegate가 정돈되지 않은 느낌을 줘서 가독성이 다소 떨어지는 느낌을 받았습니다.

커멘트 남겨드린 부분들 같이 개선해보면 좋을 것 같습니다!! 

수고 많으셨습니다!! 👏👏👏

dpwns1234 committed:
feat: ImageFileManager 기능 구현

- cache와 함께 image를 저장하도록 해 이미지 bind 속도를 올리고자 함

dpwns1234 committed:
refactor: locationManager를 viewController로 이동

- 현재 dataManager에 핵심 로직들이 다 들어있음.
- dataManager가 하는 일이 많고 복잡하므로 VC로 이동

dpwns1234 committed:
chore: 변수 이름 철자 수정

dpwns1234 committed:
fix: refresh 했을 때 데이터를 새로 받아오도록 수정

- 데이터를 새로 받아오지 않고 단순히 collection의 reload만 하고 있었던 코드 수정

dpwns1234 committed:
refactor: 매직넘버 정리

dpwns1234 committed:
refactor: ImageFileManager 싱글톤으로 리팩터

- 여러 곳에서 사용되어서 싱글톤으로 변경

dpwns1234 committed:
refactor: 이미지 아이콘을 데이터 로드할 때 이미 fileManager에 값이 있으면 네트워크 하지 않도록 리팩터

dpwns1234 committed:
refactor: IconDataService의 delegate 삭제

- 한 이미지 아이콘이 로드될 때마다 처리하는 것이 아니라 전체 다 로드 되었을 때 처리를 해주면 되기에 delegate를 없앴다. (그리고 이미 group을 통해 제어해주고 있었음)

dpwns1234 committed:
refactor: WeatherDataManager의 delegate 수정

- 어떤 데이터(today or forecast)가 로드되었는지 구별하기 위해 delegate의 함수를 completedLoadData()에서 둘로 나눴다.

dpwns1234 committed:
design: header, cell 디자인 수정

dpwns1234 committed:
fix: 로직상 문제 수정

dpwns1234 committed:
refactor: compositional layout으로 리팩터

dpwns1234 committed:
config: taget version 변경

12 -> 14.3

dpwns1234 committed:
design: 바텀 선 및 비율로 constraint 설정

dpwns1234 committed:
chore: 간단한 수정

corykim0829 commented on WeatherForecast/WeatherForecast/Controller/WeatherViewController.swift:
```diff
@@ -1,47 +1,191 @@
+
+// MARK: - Bind view
+
+extension WeatherViewController {
```
여기 bind 함수들은 뷰에게 데이터만 넘겨서 업데이트 하면 좋을 것 같습니다!

header.addressLabel.text 까지 ViewController에 오게되면 ViewController 코드가 길어지고 복잡해지니까, header나 cell에서 직접 업데이트해서 더 깔끔하게 처리하면 좋을 것 같습니다.

corykim0829 reviewed (approved):
고생많으셨습니다!

남겨드린 피드백은 다음 스텝과 함께 진행해주세요!
